src/zarr/store/zip.py

```diff
@@ -0,0 +1,221 @@
+        key: str,
+        prototype: BufferPrototype,
+        byte_range: tuple[int | None, int | None] | None = None,
+    ) -> Buffer | None:
+        try:
+            with self._zf.open(key) as f:  # will raise KeyError
+                if byte_range is None:
+                    return prototype.buffer.from_bytes(f.read())
+                start, length = byte_range
+                if start:
+                    f.seek(start or 0)
+                if length:
+                    return prototype.buffer.from_bytes(f.read(length))
+                else:
+                    return prototype.buffer.from_bytes(f.read())
+        except KeyError:
+            return None
+
+    async def get(
+        self,
+        key: str,
+        prototype: BufferPrototype,
+        byte_range: tuple[int | None, int | None] | None = None,
+    ) -> Buffer | None:
+        assert isinstance(key, str)
+
+        with self._lock:
+            return self._get(key, prototype=prototype, byte_range=byte_range)
+
+    async def get_partial_values(
+        self,
+        prototype: BufferPrototype,
+        key_ranges: list[tuple[str, tuple[int | None, int | None]]],
+    ) -> list[Buffer | None]:
+        """
+        Read byte ranges from multiple keys.
+
+        Parameters
+        ----------
+        key_ranges: List[Tuple[str, Tuple[int, int]]]
+            A list of (key, (start, length)) tuples. The first element of the tuple is the name of
+            the key in storage to fetch bytes from. The second element the tuple defines the byte
+            range to retrieve. These values are arguments to `get`, as this method wraps
+            concurrent invocation of `get`.
+        """
+        out = []
+        with self._lock:
+            for key, byte_range in key_ranges:
+                out.append(self._get(key, prototype=prototype, byte_range=byte_range))
+        return out
+
+    def _set(self, key: str, value: Buffer) -> None:
+        # generally, this should be called inside a lock
+        keyinfo = zipfile.ZipInfo(filename=key, date_time=time.localtime(time.time())[:6])
+        keyinfo.compress_type = self.compression
+        if keyinfo.filename[-1] == os.sep:
+            keyinfo.external_attr = 0o40775 << 16  # drwxrwxr-x
+            keyinfo.external_attr |= 0x10  # MS-DOS directory flag
+        else:
+            keyinfo.external_attr = 0o644 << 16  # ?rw-r--r--
+        self._zf.writestr(keyinfo, value.to_bytes())
+
+    async def set(self, key: str, value: Buffer) -> None:
+        self._check_writable()
+        assert isinstance(key, str)
+        if not isinstance(value, Buffer):
+            raise TypeError("ZipStore.set(): `value` must a Buffer instance")
+        with self._lock:
+            self._set(key, value)
+
+    async def set_partial_values(self, key_start_values: list[tuple[str, int, bytes]]) -> None:
+        raise NotImplementedError
+
+    async def delete(self, key: str) -> None:
+        raise NotImplementedError
```

Comment:
Today I was reminded that you can't delete anything from inside a ZipFile 😢. This behavior also existed in 2.18.

Reply:
Wouldn't it be nice to have an In-memory version of a Zip store where all the zip data is read in memory. This way it can support deleting and updating entries. Thereafter, a user can persist the data using a method like `write_to_file`. I think it would be very efficient for data sets whose compressed size is small enough to fit entirely in memory.

Reply:
Indeed that would be nice @zoj613. I would like to save that until after the 3.0 release though as the minimal zip store is a release blocker at this point.